flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -885,6 +850,53 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
         }
     }
 
+    /**
+     * Transition into our final state in case of failure. We should be either in DEPLOYING,
+     * INITIALIZING, RUNNING, CANCELING, or FAILED loop for multiple retries during concurrent state
```

Comment:
nit: The sentence does not read well. I suggest 
- full stop after FAILED
- then `Loop to asynchronously clean up via calls to cancel() or to failExternally()`